server/datastore/mysql/apple_mdm.go

```diff
@@ -715,28 +715,54 @@ func (ds *Datastore) IngestMDMAppleDevicesFromDEPSync(ctx context.Context, devic
 			args = append(args, d.SerialNumber)
 			parts = append(parts, "?")
 		}
-		var hosts []fleet.Host
-		err = sqlx.SelectContext(ctx, tx, &hosts, fmt.Sprintf(`
-			SELECT id, hardware_model, hardware_serial FROM hosts WHERE hardware_serial IN(%s)`,
+		var hostsWithMDMInfo []hostWithMDMInfo
+		err = sqlx.SelectContext(ctx, tx, &hostsWithMDMInfo, fmt.Sprintf(`
+			SELECT
+				h.id,
+				h.hardware_model,
+				h.hardware_serial,
+				COALESCE(hmdm.enrolled, 0) as enrolled
+			FROM hosts h
+			LEFT JOIN host_mdm hmdm ON hmdm.host_id = h.id
+			WHERE h.hardware_serial IN(%s)`,
 			strings.Join(parts, ",")),
 			args...)
 		if err != nil {
 			return ctxerr.Wrap(ctx, err, "ingest mdm apple host get host ids")
 		}
 
+		var hosts []fleet.Host
+		var unmanagedHostIDs []uint
+		for _, h := range hostsWithMDMInfo {
+			hosts = append(hosts, h.Host)
+			if h.Enrolled == nil || !*h.Enrolled {
+				unmanagedHostIDs = append(unmanagedHostIDs, h.ID)
+			}
+		}
+
 		if err := upsertMDMAppleHostDisplayNamesDB(ctx, tx, hosts...); err != nil {
 			return ctxerr.Wrap(ctx, err, "ingest mdm apple host upsert display names")
 		}
 
 		if err := upsertMDMAppleHostLabelMembershipDB(ctx, tx, ds.logger, hosts...); err != nil {
 			return ctxerr.Wrap(ctx, err, "ingest mdm apple host upsert label membership")
 		}
-
-		var ids []uint
-		for _, h := range hosts {
-			ids = append(ids, h.ID)
-		}
-		if err := upsertMDMAppleHostMDMInfoDB(ctx, tx, appCfg.ServerSettings, true, ids...); err != nil {
+		if err := upsertHostDEPAssignmentsDB(ctx, tx, hosts); err != nil {
+			return ctxerr.Wrap(ctx, err, "ingest mdm apple host upsert DEP assignments")
+		}
+
+		// only upsert MDM info for hosts that don't have any. This
+		// prevents us from overriding valuable info with potentially
+		// incorrect data. For example: if a host is enrolled in a
+		// third-party MDM, but gets assigned in ABM to Fleet (during
+		// migration) we'll get an 'added' event.
```

Comment:
my thinking was: yes because `host_mdm` is being updated by osquery for those hosts. This is sort of a hack to pre-fill MDM information for hosts that haven't been through the setup assistant yet.

Reply:
This was @roperzh's addition, but I think it's more that we can confidently skip upserting the `host_mdm` for the managed records because we can rely on osquery to ensure that value gets updated when the host is actually MDM-enrolled to Fleet.